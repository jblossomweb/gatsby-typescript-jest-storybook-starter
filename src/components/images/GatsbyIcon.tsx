import React from "react"
import { useStaticQuery, graphql } from "gatsby"
import Img, { FluidObject } from "gatsby-image"

export const query = graphql`
  query ImageGatsbyIconQuery {
    placeholderImage: file(relativePath: { eq: "gatsby-icon.png" }) {
      childImageSharp {
        fluid(maxWidth: 256) {
          aspectRatio
          src
          srcSet
          sizes
          base64
          tracedSVG
          srcWebp
          srcSetWebp
        }
      }
    }
  }
`

export interface Props {
  fluid?: FluidObject | FluidObject[]
}

export const GatsbyIcon: React.FC<Props> = ({ fluid }) => <Img fluid={fluid} />

export default () => {
  const { placeholderImage } = useStaticQuery(query)
  return <GatsbyIcon fluid={placeholderImage.childImageSharp.fluid} />
}
