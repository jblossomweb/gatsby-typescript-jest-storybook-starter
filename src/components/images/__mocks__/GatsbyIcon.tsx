import React from "react"
import Img, { FluidObject } from "gatsby-image"
import MockImageGatsbyIconQuery from "../../../../__mocks__/query/image-gatsby-icon.query.json"

export interface Props {
  fluid?: FluidObject | FluidObject[]
}

export const GatsbyIcon: React.FC<Props> = ({ fluid }) => <Img fluid={fluid} />

export default () => {
  const { placeholderImage } = MockImageGatsbyIconQuery.data
  return <GatsbyIcon fluid={placeholderImage.childImageSharp.fluid} />
}
