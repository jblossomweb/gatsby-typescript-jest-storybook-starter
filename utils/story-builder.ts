import * as knobs from "@storybook/addon-knobs"
import { storiesOf } from "@storybook/react"

const { withKnobs } = knobs

export interface KnobsInterface {
  text: (name: string, val: string) => string
  number: (name: string, val: number) => number
  date: (name: string, val: number) => number
  boolean: (name: string, val: boolean) => boolean
  object: (name: string, val: any) => any
  select: (name: string, options: any, val: any) => any
}

export type Story = (...args: any[]) => JSX.Element

export interface Stories {
  [key: string]: Story
}

const storyBuilder = (stories: Stories, storyPath: string) => {
  const builtStories = storiesOf(storyPath, module)
  builtStories.addDecorator(withKnobs as any)

  Object.keys(stories).forEach(key => {
    builtStories.add(key, () => stories[key](knobs))
  })
}

export default storyBuilder
