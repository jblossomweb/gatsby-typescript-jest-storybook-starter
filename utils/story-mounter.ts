import { Stories } from "./story-builder"
import "./enzyme-adapter"
import { shallow, mount, ReactWrapper } from "enzyme"
import mockKnobs from "../__mocks__/storybook-knobs"

export interface MountedStory {
  wrapper: ReactWrapper
  component: ReactWrapper
  props: any
  elements: { [key: string]: ReactWrapper }
}

const storyMounter = (
  stories: Stories,
  mainComponent: React.Component<any> | React.FC<any>,
  getElements?: (component: ReactWrapper) => MountedStory["elements"],
  mountShallow?: boolean
) => {
  const mountedStories: { [key: string]: MountedStory } = {}
  // tslint:disable-next-line: ban-types
  const mounter: Function = mountShallow ? shallow : mount
  Object.keys(stories).forEach((key: string) => {
    const story = stories[key]
    const wrapper: ReactWrapper = mounter(story(mockKnobs))
    const component: ReactWrapper = wrapper.find(mainComponent)
    const props = component.props()
    const elements = getElements ? getElements(component) : {}
    mountedStories[key] = {
      wrapper,
      component,
      props,
      elements,
    }
  })
  return mountedStories
}

export default storyMounter
