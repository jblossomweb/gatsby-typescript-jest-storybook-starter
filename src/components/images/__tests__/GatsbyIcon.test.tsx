import React from "react"
import * as Gatsby from "gatsby"
import "../../../../utils/enzyme-adapter"
import { shallow, ShallowWrapper } from "enzyme"
import snapshotTests from "../../../../utils/story-snapshot-tests"
import MockImageGatsbyIconQuery from "../../../../__mocks__/query/image-gatsby-icon.query.json"
import GatsbyIconWithQuery, { GatsbyIcon } from "../GatsbyIcon"
import { stories } from "./GatsbyIcon.stories"

describe("components/images/GatsbyIcon", () => {
  describe("story snapshots", () => {
    snapshotTests(stories)
  })

  describe("with query", () => {
    const { data } = MockImageGatsbyIconQuery
    const useStaticQuery = jest.spyOn(Gatsby, "useStaticQuery")
    let wrapper: ShallowWrapper

    beforeEach(() => {
      useStaticQuery.mockImplementation(() => data)
      wrapper = shallow(<GatsbyIconWithQuery />)
    })

    it("mounts the GatsbyIcon component", () => {
      const component = wrapper.find(GatsbyIcon)
      expect(component.length).toBe(1)
    })

    it("passes fluid from data.placeholderImage.childImageSharp.fluid", () => {
      const component = wrapper.find(GatsbyIcon)
      expect(component.prop("fluid")).toEqual(
        data.placeholderImage.childImageSharp.fluid
      )
    })
  })
})
