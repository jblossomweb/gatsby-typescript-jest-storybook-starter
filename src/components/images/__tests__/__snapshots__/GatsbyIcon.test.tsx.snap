// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/images/GatsbyIcon story snapshots default matches snapshot 1`] = `
<div
  className=" gatsby-image-wrapper"
  style={
    Object {
      "maxHeight": null,
      "maxWidth": null,
      "overflow": "hidden",
      "position": "relative",
    }
  }
>
  <div
    aria-hidden={true}
    style={
      Object {
        "paddingBottom": "100%",
        "width": "100%",
      }
    }
  />
  <img
    alt=""
    aria-hidden={true}
    src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABQAAAAUCAYAAACNiR0NAAAACXBIWXMAAAsSAAALEgHS3X78AAAD2klEQVQ4y3WUW2xUVRSGDyY+8MCLTxofvNNWqPY61Wgk4IvxEmI1ktpxOm2n01IKCg8gCYpGgxgkpTB3LjOlnaYNDQGaaC0kpbaVS0molWYql9SMGDWxRBrajt3nrOX6d+cMo5aTrOx99uXba61/7W14i0MGvrqikOEtDt+HvqcgeH+DI/LO+rKDnQ2lkUmZU2JmfWkkKWPdMlcp/0uxFnvqisJGum/YnSVNLxy2+2tlQ6KhKMrVuRF2LfezpyDEnsIQu3L87M6JcENhlGXNDVlbYRh3GRlqFmzXBkeM3Xkh3vTSEdWxe1D9OJy0bl6bopvXp2j83C/Utfd7tXl1TLlzg9zkOMreknBLFtTIhAnYprI4V+X5zOjOfnM+pYiZCbCRvut0sfcaJX/6U49ZpsXxLwbNqqd9qqm0netLwr502jJeroVngPVELlmyiS/3T9KONztJNrFz+X5trjwfb3+jg85/fRVg7u+6QrJXpT11GlkCJBAmPMPCk8ERqnxqP7tX+HE6ixgLJv3qlX6ueHwfH2s+R5ZF/PFbXWZtfhhzSYEvM6AmBGh68bD6e07R6MDP5BSYNw2SRf8yjEloIlSQt73aTmjh5YbSVrQeA6UBNeVEhfxsfz1O8EwO0huzYfa/9lqs5pkAWrKBMt9joM6qcgMc+/Qsckd9bT/oXGlIyV1Idlv7bFAbvJOWxKz1JUcwPw5lFMJY92gzdXw5pJN9Oj7GrtwDi0LhmZRUxj5YHaP3V0WpvlhHNGNIwZqNzx/i0NY+qnyyhdp3fceLQQFyiyAflXfy9NQsp2bneWY6xTO3UzQx8ivVFeq0zBm4TrgBKNqhkxNU/uAeju8e/B8UUSDMzWti3N1ynk/4LyLvfOevOZq88odVu1JXQQKidEMUgWhRvm0dpbcf3rs4VLxE3t6T/4rH9rHkn01lUU/4kgmGCPkNysaJu7nl5VY1fWuWbWHuBYX6jc8dZEkPD3SP65x/+Fpc1Rfo+90IUZbiouNuiii6sO8JFfVr8gO6sGOf9GvYcd8F05UTgLe/CesB++pV4PpI0tWZjjGyob2to1z+0FdsC4VDRFE+e2zBs+FTE1SdH5jf6GiDcF4btvA4yKux0dGuoZ17hs07t1MaOnQiQe8+0ULiqX4Y0sbHD1wwF2Bx5DaaecIyj6J8eDXgaZWEv/WVNoXbc2Psd+voZwO07pFmavt8wOqNXTYxhzDhmQ3775u4JGvQKTlNQigoh3JoLDukFa1ZEdZjWgDJmaz1LgILZR5He7CuMLRMoB4pqR7xYFz+Z1C0UosJlAbUtAXAHnsfGP8An71yIvslifEAAAAASUVORK5CYII="
    style={
      Object {
        "height": "100%",
        "left": 0,
        "objectFit": "cover",
        "objectPosition": "center",
        "opacity": 1,
        "position": "absolute",
        "top": 0,
        "transitionDelay": "500ms",
        "width": "100%",
      }
    }
  />
  <img
    alt=""
    aria-hidden={true}
    src="data:image/svg+xml,%3csvg%20xmlns='http://www.w3.org/2000/svg'%20width='400'%20height='400'%20viewBox='0%200%20400%20400'%20preserveAspectRatio='none'%3e%3cpath%20d='M173%201A201%20201%200%200023%20294a200%20200%200%2000377-78c1%201%201-7%201-17l-1-15-2-9A201%20201%200%2000173%201m5%2043A158%20158%200%200047%20164c0%201%2042%2044%2094%2095l94%2094%208-2a158%20158%200%2000114-148v-3H257v28h34l34%201a127%20127%200%2001-72%2087l-10%205-82-82-82-83a128%20128%200%2001221-36l5%205%2011-8%2010-9c2-2-25-29-37-38-18-11-37-20-57-24-11-2-44-4-54-2M43%20212a156%20156%200%2000148%20144h6l-76-76-77-77-1%209'%20fill='%23d3d3d3'%20fill-rule='evenodd'/%3e%3c/svg%3e"
    style={
      Object {
        "height": "100%",
        "left": 0,
        "objectFit": "cover",
        "objectPosition": "center",
        "opacity": 1,
        "position": "absolute",
        "top": 0,
        "transitionDelay": "500ms",
        "width": "100%",
      }
    }
  />
  <picture>
    <source
      sizes="(max-width: 256px) 100vw, 256px"
      srcSet="/static/4a9773549091c227cd2eb82ccd9c5e3a/85647/gatsby-icon.webp 64w,
/static/4a9773549091c227cd2eb82ccd9c5e3a/cde37/gatsby-icon.webp 128w,
/static/4a9773549091c227cd2eb82ccd9c5e3a/34b62/gatsby-icon.webp 256w,
/static/4a9773549091c227cd2eb82ccd9c5e3a/faab5/gatsby-icon.webp 384w,
/static/4a9773549091c227cd2eb82ccd9c5e3a/a3e81/gatsby-icon.webp 512w"
      type="image/webp"
    />
    <source
      sizes="(max-width: 256px) 100vw, 256px"
      srcSet="/static/4a9773549091c227cd2eb82ccd9c5e3a/49ddd/gatsby-icon.png 64w,
/static/4a9773549091c227cd2eb82ccd9c5e3a/62915/gatsby-icon.png 128w,
/static/4a9773549091c227cd2eb82ccd9c5e3a/61fd6/gatsby-icon.png 256w,
/static/4a9773549091c227cd2eb82ccd9c5e3a/450e9/gatsby-icon.png 384w,
/static/4a9773549091c227cd2eb82ccd9c5e3a/bc59e/gatsby-icon.png 512w"
    />
    <img
      alt=""
      loading="lazy"
      onLoad={[Function]}
      sizes="(max-width: 256px) 100vw, 256px"
      src="/static/4a9773549091c227cd2eb82ccd9c5e3a/61fd6/gatsby-icon.png"
      srcSet="/static/4a9773549091c227cd2eb82ccd9c5e3a/49ddd/gatsby-icon.png 64w,
/static/4a9773549091c227cd2eb82ccd9c5e3a/62915/gatsby-icon.png 128w,
/static/4a9773549091c227cd2eb82ccd9c5e3a/61fd6/gatsby-icon.png 256w,
/static/4a9773549091c227cd2eb82ccd9c5e3a/450e9/gatsby-icon.png 384w,
/static/4a9773549091c227cd2eb82ccd9c5e3a/bc59e/gatsby-icon.png 512w"
      style={
        Object {
          "height": "100%",
          "left": 0,
          "objectFit": "cover",
          "objectPosition": "center",
          "opacity": 0,
          "position": "absolute",
          "top": 0,
          "transition": "opacity 500ms",
          "width": "100%",
        }
      }
    />
  </picture>
  <noscript
    dangerouslySetInnerHTML={
      Object {
        "__html": "<picture><source type='image/webp' srcset=\\"/static/4a9773549091c227cd2eb82ccd9c5e3a/85647/gatsby-icon.webp 64w,
/static/4a9773549091c227cd2eb82ccd9c5e3a/cde37/gatsby-icon.webp 128w,
/static/4a9773549091c227cd2eb82ccd9c5e3a/34b62/gatsby-icon.webp 256w,
/static/4a9773549091c227cd2eb82ccd9c5e3a/faab5/gatsby-icon.webp 384w,
/static/4a9773549091c227cd2eb82ccd9c5e3a/a3e81/gatsby-icon.webp 512w\\" sizes=\\"(max-width: 256px) 100vw, 256px\\" /><source srcset=\\"/static/4a9773549091c227cd2eb82ccd9c5e3a/49ddd/gatsby-icon.png 64w,
/static/4a9773549091c227cd2eb82ccd9c5e3a/62915/gatsby-icon.png 128w,
/static/4a9773549091c227cd2eb82ccd9c5e3a/61fd6/gatsby-icon.png 256w,
/static/4a9773549091c227cd2eb82ccd9c5e3a/450e9/gatsby-icon.png 384w,
/static/4a9773549091c227cd2eb82ccd9c5e3a/bc59e/gatsby-icon.png 512w\\" sizes=\\"(max-width: 256px) 100vw, 256px\\" /><img loading=\\"lazy\\" sizes=\\"(max-width: 256px) 100vw, 256px\\" srcset=\\"/static/4a9773549091c227cd2eb82ccd9c5e3a/49ddd/gatsby-icon.png 64w,
/static/4a9773549091c227cd2eb82ccd9c5e3a/62915/gatsby-icon.png 128w,
/static/4a9773549091c227cd2eb82ccd9c5e3a/61fd6/gatsby-icon.png 256w,
/static/4a9773549091c227cd2eb82ccd9c5e3a/450e9/gatsby-icon.png 384w,
/static/4a9773549091c227cd2eb82ccd9c5e3a/bc59e/gatsby-icon.png 512w\\" src=\\"/static/4a9773549091c227cd2eb82ccd9c5e3a/61fd6/gatsby-icon.png\\" alt=\\"\\" style=\\"position:absolute;top:0;left:0;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center\\"/></picture>",
      }
    }
  />
</div>
`;
